src/core2cuda.rkt

```diff
@@ -0,0 +1,152 @@
+  (supported-list
+    (invert-op-proc (curry set-member? '(array dim size ref for for* tensor tensor*)))
+    fpcore-consts
+    (curry set-member? '(binary32 binary64 binary80 integer))
+    (invert-rnd-mode-proc (curry equal? 'nearestAway))
+    #f))
+
+(define (default-unknown->cuda ctx op args)
+  (define type (type->cuda (ctx-lookup-prop ctx ':precision)))
+  (format "~a~a(~a)"
+          op
+          (cuda-type->suffix type)
+          (string-join args ", ")))
+
+(define unknown->cuda (make-parameter default-unknown->cuda))
+
+(define (set-cuda-header! gen-proc)
+  (set! cuda-header (gen-proc cuda-header)))
+
+(define (set-unknown->cuda! gen-proc)
+  (unknown->cuda (gen-proc (unknown->cuda))))
+
+(define cuda-reserved  ; Language-specific reserved names (avoid name collisions)
+  '(auto break case char const continue default
+    do double else enum extern float for goto if
+    inline int long register restrict return
+    short signed sizeof static struct switch
+    typedef union unsigned void volatile while))
+
+(define/match (cuda-type->suffix type)
+  [("int64_t") ""]
+  [("double") ""]
+  [("float") "f"]
+  [("long double") "l"])
+
+(define/match (type->cuda type)
+  [('binary64) "double"]
+  [('binary32) "float"]
+  [('binary80) "long double"]
+  [('boolean) "int"]
+  [('integer) "int64_t"])
+
+(define (operator->cuda op args ctx)
+  (define type (type->cuda (ctx-lookup-prop ctx ':precision)))
+  (match op
+   [(or 'isinf 'isnan 'isfinite 'isnormal 'signbit)
+    (format "~a~a" op args)]
+   [_
+    ((unknown->cuda) ctx op args)]))
+  
+(define (binary80->string x)
+  (parameterize ([gfl-exponent 15] [gfl-bits 80])
+    (let ([s (gfl->string (gfl x))])
+      (if (string-contains? s ".") s (string-append s ".0")))))
+
+(define (constant->cuda x ctx)
+  (define type (type->cuda (ctx-lookup-prop ctx ':precision)))
+  (match x
+   [(or 'TRUE 'FALSE (? hex?)) (~a x)]
+   [(or 'M_1_PI 'M_2_PI 'M_2_SQRTPI 'INFINITY 'NAN)
+    (format "((~a) ~a)" type x)]
+   [(? number?)
+    (match type
+     ["int64_t" (~a (inexact->exact x))]
+     ["long double" (format "~a~a" (binary80->string x) (cuda-type->suffix type))]
+     [_ (format "~a~a" (real->double-flonum x) (cuda-type->suffix type))])]
+   [(? symbol?) (format "((~a) M_~a)" type x)]))
+  
+(define (round->cuda x ctx)
+  (define type (type->cuda (ctx-lookup-prop ctx ':precision)))
+  (format "((~a) ~a)" type x))
+
+(define (cmp-prec prec1 prec2)
+  (define/match (prec->num prec)
+    [('binary80) 4]
+    [('binary64) 3]
+    [('binary32) 2]
+    [('integer)  1]
+    [('boolean)  0])
+  (- (prec->num prec1) (prec->num prec2)))
+
+(define (implicit-round->cuda op arg arg-ctx ctx)
+  (define prec (ctx-lookup-prop ctx ':precision))
+  (define arg-prec (ctx-lookup-prop arg-ctx ':precision))
+  (if (set-member? '(+ - * /) op)
+      (if (> (cmp-prec prec arg-prec) 0)
+          (round->cuda arg ctx)
+          arg)  ; TODO: warn unfaithful
+      arg))
+
+(define (round-mode->cuda mode ctx)
```

Comment:
I don't think CUDA supports rounding modes. It seems to support intrinsics for certain operations that have the rounding mode built in. That would be cool to support but maybe not interesting to you now—but we shouldn't have the wrong code here.

Reply:
Ok removed